### PR TITLE
Add NginxHostedApp deploy hook

### PR DESCRIPTION
**Pull request by serhii-karelov**
Move out apache hooks from `DjangoApp` to new `ApacheHostedDjangoApp`. 
Rename `DjangoMultiSiteApp` into `ApacheHostedDjangoMultiSiteApp`
Add `NginxHostedApp` and `UwsgiHostedApp` deploy hooks 

@yola/service-engineering @yola/operations-engineering

### CODE REVIEW

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by RayeN** [comment]
tests?

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by serhii-karelov** [comment]
> tests?

Fixed

File: yodeploy/hooks/tests/test_django.py [hunk 1]
```diff
@@ -3,13 +3,14 @@
 from mock import PropertyMock, call, patch
 from pkg_resources import parse_version
 
-from yodeploy.hooks.django import DjangoApp
+from yodeploy.hooks.django import ApacheHostedDjangoApp, DjangoApp
 
 
 class TestDjangoHook(unittest.TestCase):
```
Comment (jamesodo) [inline yodeploy/hooks/tests/test_django.py]: Rename `TestDjangoHook` to `TestApacheHostedDjangoAppHook`?
I'm assuming the current `test_vhost_attributes ` would fail for `NginxHostedDjangoApp`.

File: yodeploy/hooks/tests/test_nginx.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    def test_places_logfiles(self):
+        self.dh.prepare()
+
+        self.touch_mock.assert_has_calls([
+            call('/var/log/nginx/nginx-app-access.log',
+                 'www-data', 'adm', 0o640),
+            call('/var/log/nginx/nginx-app-error.log',
+                 'www-data', 'adm', 0o640),
+            ],
```
Comment (snitch) [inline yodeploy/hooks/tests/test_nginx.py]: E123 closing bracket does not match indentation of opening bracket's line

File: yodeploy/test_integration/samples/apache-hosted-django-app/deploy/hooks.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+class Hooks(ApacheHostedDjangoApp):
+
+    # these overrides are only needed for testsing
+    apache = mock.Mock()
+    vhost_path = join(tests_dir, 'filesys', 'etc', 'apache2', 'sites-enabled')
+    includes_path = join(tests_dir, 'filesys', 'etc', 'apache2', 'yola.d')
+    log_user = getpass.getuser()
+    log_group = None
+
+hooks = Hooks
```
Comment (snitch) [inline yodeploy/test_integration/samples/apache-hosted-django-app/deploy/hooks.py]: E305 expected 2 blank lines after class or function definition, found 1

File: yodeploy/test_integration/samples/nginx-app/deploy/hooks.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+class Hooks(NginxHostedApp):
+    # these overrides are only needed for testing
+    nginx = mock.Mock()
+    logs_path = join(tests_dir, 'filesys/var/log/nginx')
+    server_blocks_path = join(tests_dir, 'filesys/etc/nginx/sites-enabled')
+    user = getpass.getuser()
+    group = None
+
+hooks = Hooks
```
Comment (snitch) [inline yodeploy/test_integration/samples/nginx-app/deploy/hooks.py]: E305 expected 2 blank lines after class or function definition, found 1

File: yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+class Hooks(UwsgiHostedApp):
+    # these overrides are only needed for testing
+    uwsgi = mock.Mock()
+    logs_path = join(tests_dir, 'filesys/var/log/uwsgi')
+    config_path = join(tests_dir, 'filesys/etc/uwsgi')
+    user = getpass.getuser()
+    group = None
+
+hooks = Hooks
```
Comment (snitch) [inline yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: E305 expected 2 blank lines after class or function definition, found 1

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: yodeploy/hooks/nginx.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import subprocess
+import sys
+
+from yodeploy.hooks.configurator import ConfiguratedApp
+from yodeploy.util import touch
+
+log = logging.getLogger(__name__)
+
+
+class Nginx(object):
```
Comment (ekoval) [inline yodeploy/hooks/nginx.py]: I'm not sure it worth a separate class. Maybe just add `_reload_nginx` method to the hook.

File: yodeploy/hooks/nginx.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    @staticmethod
+    def reload():
+        try:
+            subprocess.check_call(('service', 'nginx', 'reload'))
+        except subprocess.CalledProcessError:
```
Comment (ekoval) [inline yodeploy/hooks/nginx.py]: It would have been nice to see more details here. Like this:
```
try:
   subprocess.check_output(cmd, stderr=subprocess.STDOUT)
except subprocess.CalledProcessError as e:
    log(e.returncode, e.output)
```

File: yodeploy/test_integration/samples/nginx-app/deploy/hooks.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+from yodeploy.hooks.nginx import NginxHostedApp
+from yodeploy.test_integration.helpers import tests_dir
+
+
+class Hooks(NginxHostedApp):
+    # these overrides are only needed for testing
+    nginx = mock.Mock()
+    logs_path = join(tests_dir, 'filesys/var/log/nginx')
+    server_blocks_path = join(tests_dir, 'filesys/etc/nginx/sites-enabled')
```
Comment (ekoval) [inline yodeploy/test_integration/samples/nginx-app/deploy/hooks.py]: Why it is called this way? Not "enabled_sites"?

File: yodeploy/test_integration/samples/nginx-app/deploy/hooks.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+from os.path import join
+
+from yodeploy.hooks.nginx import NginxHostedApp
+from yodeploy.test_integration.helpers import tests_dir
+
+
+class Hooks(NginxHostedApp):
+    # these overrides are only needed for testing
+    nginx = mock.Mock()
+    logs_path = join(tests_dir, 'filesys/var/log/nginx')
```
Comment (ekoval) [inline yodeploy/test_integration/samples/nginx-app/deploy/hooks.py]: I think you better use `os.path.join` instead.

File: yodeploy/test_integration/samples/nginx-app/deploy/hooks.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import mock
+
+from os.path import join
+
+from yodeploy.hooks.nginx import NginxHostedApp
+from yodeploy.test_integration.helpers import tests_dir
+
+
+class Hooks(NginxHostedApp):
+    # these overrides are only needed for testing
```
Comment (ekoval) [inline yodeploy/test_integration/samples/nginx-app/deploy/hooks.py]: "overridE" - typo

File: yodeploy/test_integration/samples/nginx-app/deploy/hooks.py (hunk 9 above)
Comment (serhii-karelov) [reply yodeploy/test_integration/samples/nginx-app/deploy/hooks.py]: It is the same function.

File: yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
+import getpass
+
```
Comment (ekoval) [inline yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: Why do we need these two empty lines?

File: yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+from os.path import join
+
+from yodeploy.hooks.uwsgi import UwsgiHostedApp
+from yodeploy.test_integration.helpers import tests_dir
+
+
+class Hooks(UwsgiHostedApp):
+    # these overrides are only needed for testing
+    uwsgi = mock.Mock()
+    logs_path = join(tests_dir, 'filesys/var/log/uwsgi')
```
Comment (ekoval) [inline yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: `os.path.join`

File: yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+import mock
+
+from os.path import join
+
+from yodeploy.hooks.uwsgi import UwsgiHostedApp
+from yodeploy.test_integration.helpers import tests_dir
+
+
+class Hooks(UwsgiHostedApp):
+    # these overrides are only needed for testing
```
Comment (ekoval) [inline yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: Probably needless comment here.

File: yodeploy/test_integration/samples/nginx-app/deploy/hooks.py (hunk 9 above)
Comment (ekoval) [reply yodeploy/test_integration/samples/nginx-app/deploy/hooks.py]: Oh, it is imported from `os`. I wouldn't to do this, it is easy to mix up with string `''.join()`

File: yodeploy/hooks/nginx.py (hunk 6 above)
Comment (serhii-karelov) [reply yodeploy/hooks/nginx.py]: I want to be consistent with existing code (Apache hooks for example)

File: yodeploy/hooks/nginx.py (hunk 7 above)
Comment (serhii-karelov) [reply yodeploy/hooks/nginx.py]: +1

File: yodeploy/hooks/nginx.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+    def deployed(self):
+        super(NginxHostedApp, self).deployed()
+        self.nginx_hosted_deployed()
+
+    def prepare(self):
+        super(NginxHostedApp, self).prepare()
+        self.nginx_hosted_prepare()
+
+    def nginx_hosted_prepare(self):
```
[the remark this answers is not in the document]
Comment (serhii-karelov) [reply yodeploy/hooks/nginx.py]: I try to write code with a consistent style, check out any other hook

File: yodeploy/test_integration/samples/nginx-app/deploy/hooks.py (hunk 8 above)
Comment (serhii-karelov) [reply yodeploy/test_integration/samples/nginx-app/deploy/hooks.py]: It is nginx terminology. https://www.nginx.com/resources/wiki/start/topics/examples/server_blocks/
So I've decided to name it server block instead of vhost (which is Apache terminology).

File: yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py (hunk 11 above)
Comment (serhii-karelov) [reply yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: i'll reorder the imports

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,6 +1,14 @@
 Change Log
 ==========
 
+0.9.0
+-----
+
+* Move out apache hooks from `DjangoApp` to new `ApacheHostedDjangoApp`;
```
Comment (RayeN) [inline CHANGELOG.md]: we usually use `.` ;)

File: yodeploy/hooks/nginx.py (hunk 6 above)
Comment (serhii-karelov) [reply yodeploy/hooks/nginx.py]: And it is easier to plug mock for testing

File: yodeploy/hooks/uwsgi.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        )
+
+    def uwsgi_hosted_deployed(self):
+        log.debug('Running UwsgiHostedApp deployed hook.')
+        self._reload_uwsgi()
+
+    def _reload_uwsgi(self):
+        try:
+            subprocess.check_call(('service', 'uwsgi', 'reload'))
+        except subprocess.CalledProcessError as e:
```
Comment (snitch) [inline yodeploy/hooks/uwsgi.py]: F841 local variable 'e' is assigned to but never used

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: yodeploy/hooks/django.py [hunk 17]
```diff
@@ -145,6 +143,21 @@ def manage_py(self, command, *args):
...
 
 
-class DjangoMultiSiteApp(ApacheMultiSiteApp, DjangoApp):
+class ApacheHostedDjangoApp(DjangoApp, ApacheHostedApp):
+    def prepare(self):
+        super(ApacheHostedDjangoApp, self).prepare()
+        if self.template_exists('apache2/wsgi-handler.wsgi.template'):
+            self.template(
+                'apache2/wsgi-handler.wsgi.template',
+                self.deploy_path(self.app + '.wsgi'))
```
[the remark this answers is not in the document]
Comment (serhii-karelov) [reply yodeploy/hooks/django.py]: @RayeN +1

@yola/operations-engineering 
I'm going to remove this piece. Looks like it is legacy.

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by serhii-karelov** [comment]
@yola/operations-engineering please review.

Code that renders `wsgi-handler.wsgi.template`  has been removed.

File: yodeploy/hooks/nginx.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ekoval) [reply yodeploy/hooks/nginx.py]: I can understand following existing conventions, but not for private methods. I don't think you'll break any rules if you make this "private".

[the remark this answers is not in the document]
Comment (jamesodo) [reply yodeploy/hooks/nginx.py]: It might be a good idea to keep it public just in case it ever needs to be overridden.

File: yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+    logs_path = os.path.join(tests_dir, 'filesys/var/log/uwsgi')
+    config_path = os.path.join(tests_dir, 'filesys/etc/uwsgi')
+    user = getpass.getuser()
+    group = None
+
+    _reload_uwsgi = Mock()
```
Comment (beck) [inline yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: Thinking of tests as the first consumers of the code, if short circuiting uwsgi is useful then the method should probably be public, not private.

File: yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py (hunk 13 above)
Comment (beck) [reply yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: Personal pref is to always bake comments into readable code:
```diff
- class Hooks(UwsgiHostedApp):
+ class HooksWithTestSetup(UwsgiHostedApp):
```

Comment (serhii-karelov) [reply yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: > Personal pref is to always bake comments into readable code:
> ```diff
> - class Hooks(UwsgiHostedApp):
> + class HooksWithTestSetup(UwsgiHostedApp):
>```

That indeed looks better!

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py (hunk 18 above)
Comment (RayeN) [reply yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: Slightly disagree here.  IMO tests shouldn't prevent us from making method/function protected/private.  Otherwise, we won't have protected/private functions/methos at all.

Comment (beck) [reply yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: I heard one opinion (which I think is slowly starting to become my own) that if your language doesn't support public/private members, don't try to pretend like that feature exists. Because even if your member is private, somewhere/someone/something will consume it (as we see in this test).

So embrace it. Make it public and don't worry about 🙈🙉🎉

Comment (RayeN) [reply yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: Yeah, from technical point of view - we don't have private methods :)  But it's kinda convenient to say all other developers about your intention.  IMO this helps a lot to understand the code.  Thus let me rephrase your proposal a bit:  don't worry about mocking/testing so-called private/protected methods/functions in python :)

Comment (ekoval) [reply yodeploy/test_integration/samples/uwsgi-app/deploy/hooks.py]: Also, technically variable names don't matter at all. Which doesn't mean we shouldn't spent hours thinking of them.

File: yodeploy/hooks/nginx.py [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+        log.debug('Running NginxHostedApp prepare hook.')
+        access_log_path = os.path.join(
+            self.logs_path, '{}-access.log'.format(self.app)
+        )
+        error_log_path = os.path.join(
+            self.logs_path, '{}-error.log'.format(self.app)
+        )
+        touch(access_log_path, self.user, self.group, 0o640)
+        touch(error_log_path, self.user, self.group, 0o640)
```
Comment (stefanor) [inline yodeploy/hooks/nginx.py]: Don't think you need to - nginx has write access to this directory, and will create them itself.

File: yodeploy/hooks/uwsgi.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+            os.path.join(self.config_path, '{}.ini'.format(self.app)),
+        )
+
+    def uwsgi_hosted_deployed(self):
+        log.debug('Running UwsgiHostedApp deployed hook.')
+        self.reload_uwsgi()
+
+    def reload_uwsgi(self):
+        try:
+            subprocess.check_call(('service', 'uwsgi', 'reload'))
```
Comment (stefanor) [inline yodeploy/hooks/uwsgi.py]: We have been running separate services for each app, There is no uwsgi service.

Comment (serhii-karelov) [reply yodeploy/hooks/uwsgi.py]: @stefanor
Yes, I missed that. 
Is it ok if I get rid of UwsgiHostedApp for now?
Starting/restarting of uwsgi will be handled by DaemonApp. And config placement will be handled in each app hooks separately.

Comment (stefanor) [reply yodeploy/hooks/uwsgi.py]: :+1:

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by serhii-karelov** [comment]
@stefanor removed `UwsgiHostedApp`

**Comment by serhii-karelov** [comment]
@stefanor please check it once more. I.ve updated it.

Thanks
